File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0382_Linked_List_Random_Node.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Random;

public class Question_0382_Linked_List_Random_Node {
    // Definition for singly-linked list.
    public class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    private ListNode head;
    private Random random;

    /** @param head The linked list's head.
    Note that the head is guaranteed to be not null, so it contains at least one node. */
    public Question_0382_Linked_List_Random_Node(ListNode head) {
        this.head = head;
        this.random = new Random();
    }

    /** Returns a random node's value. */
    public int getRandom() {
        int res = head.val, i = 2;
        ListNode cur = head.next;
        while (cur != null) {
            int j = random.nextInt() % i;
            if (j == 0) {
                res = cur.val;
            }
            ++i;
            cur = cur.next;
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0382_Linked_List_Random_Node.cc
#include <stdlib.h>

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
 public:
  /** @param head The linked list's head.
    Note that the head is guaranteed to be not null, so it contains at least one node. */
  Solution(ListNode* head) {
    h = head;
  }
    
  /** Returns a random node's value. */
  int getRandom() {
    int res = h->val, i = 2;
    ListNode *cur = h->next;
    while (cur) {
      int j = rand() % i;
      if (j == 0) {
        res = cur->val;
      }
      ++i;
      cur = cur->next;
    }
    return res;
  }

 private:
  ListNode *h;
};
